### DISCUSSION TITLE
[Next 14.2.5] Using Next.js Data Cache to Persist File System Reads Across Server Requests

### DISCUSSION THREAD

**Discussion by CharlyMartin** [vercel/next.js#73512] [Help]
### Summary

I'm trying to understand how we can use Next's built-in [Data Cache](https://nextjs.org/docs/14/app/building-your-application/caching#data-cache) to **persist the result of file system reads across server requests.** We are on Next.js 14.2.5.

We perform these file system reads on pages using server-side search params, which is a [dynamic API](https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-apis). This opts the pages into dynamic rendering, hence our need for caching.

The examples I see in the documentation all mention using `fetch` with a `revalidate` option, but we're not using `fetch`, we're using the `fs.readFileSync` to read Markdown files and populate our pages with their content.

I did see this section about [Fetching data on the Server with third-party libraries](https://nextjs.org/docs/14/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-server-with-third-party-libraries).

> In cases where you're using a third-party library that doesn't support or expose fetch (for example, a database, CMS, or ORM client), you can configure the caching and revalidating behavior of those requests using the Route Segment Config Option and React's cache function.

`fs.readFileSync` falls under that category as it doesn't expose fetch. So, is wrapping React's `cache` around `readFileSync` the right approach for our use case?

I also read in this [section](https://nextjs.org/docs/14/app/building-your-application/caching#request-memoization) that _For cases where fetch is not suitable ... you can use the React cache function to memoize functions._ But memoization only lasts for the duration of a single request. And we want content caching across requests.

Below is a simplified example of what we're doing. In most cases, we would only want to read files' content once during the build, cache it, and only read again during the following build, as the content doesn't change in between.

```ts
// src/utils/getFrontmatterContent.ts
import fs from "fs"

import matter from "gray-matter"

export function getFrontmatterContent(path: string) {
  const fileContent = fs.readFileSync(path, "utf8")
  const { data: frontmatter } = matter(fileContent)

  return frontmatter
}
```

```tsx
// src/app/blog/page.tsx
import { getFrontmatterContent } from "@/utils/getFrontmatterContent"

// We want content to be cached across requests
const content = getFrontmatterContent("src/content/pages/blog.md")

export default async function Blog({ searchParams }) {
  return (
    <PageLayout>
      <h1>{content.title}</h1>
      <div>{content.description}</div>
    </PageLayout>
  )
}
```

**Comment by xarthurx** [comment]
Same question here, did you resolve it?
